Mathlib/RingTheory/AdicCompletion/Basic.lean

```diff
@@ -201,67 +203,228 @@ instance (priority := 100) of_subsingleton [Subsingleton M] : IsPrecomplete I M
+#align adic_completion.coe_eval AdicCompletion.coe_eval
 
-theorem eval_apply (n : ℕ) (f : adicCompletion I M) : eval I M n f = f.1 n :=
+theorem eval_apply (n : ℕ) (f : AdicCompletion I M) : eval I M n f = f.1 n :=
   rfl
-#align adic_completion.eval_apply adicCompletion.eval_apply
+#align adic_completion.eval_apply AdicCompletion.eval_apply
 
-theorem eval_of (n : ℕ) (x : M) : eval I M n (of I M x) = mkQ _ x :=
+theorem eval_of (n : ℕ) (x : M) : eval I M n (of I M x) = mkQ (I ^ n • ⊤ : Submodule R M) x :=
   rfl
-#align adic_completion.eval_of adicCompletion.eval_of
+#align adic_completion.eval_of AdicCompletion.eval_of
 
 @[simp]
 theorem eval_comp_of (n : ℕ) : (eval I M n).comp (of I M) = mkQ _ :=
   rfl
-#align adic_completion.eval_comp_of adicCompletion.eval_comp_of
+#align adic_completion.eval_comp_of AdicCompletion.eval_comp_of
 
 @[simp]
 theorem range_eval (n : ℕ) : LinearMap.range (eval I M n) = ⊤ :=
   LinearMap.range_eq_top.2 fun x => Quotient.inductionOn' x fun x => ⟨of I M x, rfl⟩
-#align adic_completion.range_eval adicCompletion.range_eval
+#align adic_completion.range_eval AdicCompletion.range_eval
 
 variable {I M}
 
 @[ext]
-theorem ext {x y : adicCompletion I M} (h : ∀ n, eval I M n x = eval I M n y) : x = y :=
+theorem ext {x y : AdicCompletion I M} (h : ∀ n, eval I M n x = eval I M n y) : x = y :=
   Subtype.eq <| funext h
-#align adic_completion.ext adicCompletion.ext
+#align adic_completion.ext AdicCompletion.ext
 
 variable (I M)
 
-instance : IsHausdorff I (adicCompletion I M) :=
+instance : IsHausdorff I (AdicCompletion I M) :=
   ⟨fun x hx => ext fun n => smul_induction_on (SModEq.zero.1 <| hx n) (fun r hr x _ =>
     ((eval I M n).map_smul r x).symm ▸
       Quotient.inductionOn' (eval I M n x) fun x => SModEq.zero.2 <| smul_mem_smul hr mem_top)
     fun _ _ ih1 ih2 => by rw [LinearMap.map_add, ih1, ih2, LinearMap.map_zero, add_zero]⟩
 
-end adicCompletion
+@[simp]
+theorem transitionMap_mk {m n : ℕ} (hmn : m ≤ n) (x : M) :
+    transitionMap I M hmn
+      (Submodule.Quotient.mk (p := (I ^ n • ⊤ : Submodule R M)) x) =
+      Submodule.Quotient.mk (p := (I ^ m • ⊤ : Submodule R M)) x := by
+  simp [transitionMap]
+
+@[simp]
+theorem transitionMap_eq (n : ℕ) : transitionMap I M (Nat.le_refl n) = LinearMap.id := by
+  ext
+  simp
+
+@[simp]
+theorem transitionMap_comp {m n k : ℕ} (hmn : m ≤ n) (hnk : n ≤ k) :
+    transitionMap I M hmn ∘ₗ transitionMap I M hnk = transitionMap I M (hmn.trans hnk) := by
+  ext
+  simp
+
+@[simp]
+theorem transitionMap_comp_apply {m n k : ℕ} (hmn : m ≤ n) (hnk : n ≤ k)
+    (x : M ⧸ (I ^ k • ⊤ : Submodule R M)) :
+    transitionMap I M hmn (transitionMap I M hnk x) = transitionMap I M (hmn.trans hnk) x := by
+  change (transitionMap I M hmn ∘ₗ transitionMap I M hnk) x = transitionMap I M (hmn.trans hnk) x
+  simp
+
+@[simp]
+theorem transitionMap_comp_eval_apply {m n : ℕ} (hmn : m ≤ n) (x : AdicCompletion I M) :
+    transitionMap I M hmn (x.val n) = x.val m :=
```

Comment:
Which is the simp normal form? `eval I M n f` or `f.1 n`? You should choose one and design the simp lemmas around it. If it is the former, than this lemma should be stated with it. If it is the latter, `AdicCompletion.eval_apply` should be a simp lemma and `val_add`, `val_sub`, `val_smul`, `val_zero` simp lemmas should be added.

Reply:
The simp NF is `f.1 n` because of `AdicCompletion.coe_eval` being a simp lemma. That's why `AdicCompletion.eval_apply` can't be a simp lemma currently, because it is already proven by `simp`. Are you saying that `eval_apply` should be the simp lemma instead of `coe_eval`?

For the time being, I added the simp lemmas you suggested. Anyway, I am not entirely sure whether the simp NF should not be `eval I M n f` instead, because it unveils less of the internal implementation of `AdicCompletion`. It might also be easier to control in the `Algebra` case, since then there are two "canonical" projections. The one to `R ⧸ (I ^ n • ⊤)` (`eval I R n`) and the one to `R ⧸ I ^ n` (`evalₐ I n` which is defined in the Algebra PR).